Add user_action app for gathering telemetry on user actions

cjcenizal opened the pull request:
## In this PR

Fixes https://github.com/elastic/kibana/issues/31330

- Adds an OSS app called `user_action`, which publishes a single API endpoint: `POST /user_action/<app>/<action_type>`
- Updates the Rollup Job Wizard to hit this endpoint when the tries to create a rollup job
- Updates the Rollup Job Wizard's usage collector to publish this metric

In the collector, this takes the form of:

```json
{
  "user_actions": {
    "create_rollup_job": 4
  }
}
```

## To-do

- [x] Add API integration tests
- [ ] ~~Per feedback from Jon B., throttle requests to one every 5 seconds~~ Unlike sending an API request per keystroke as someone types, these actions will be dispatched when users click buttons and tabs, which is unlikely to exceed a request every few seconds.

## Testing

1. Create a rollup job (Management > Rollup Jobs) with no "Delay", and a "Frequency" of "Every minute".
2. Visit `http://localhost:5601/api/stats?extended=true` and look for these stats to reflect the saved objects you've created.

To see all of the generated user action saved objects, go to Dev Tools and run this:

```
GET _search
{
  "query": {
    "match": {
      "type": "user-action"
    }
  }
}
```


# Dev Docs
## Add user_action app for gathering telemetry on user actions
Added a new API endpoint `POST /user_action/<app>/<action_type>` which is used to add UI telemetry to track user actions. This API that is not intended for external use and will likely change in the next versions.

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
Pinging @elastic/kibana-platform

cjcenizal commented on x-pack/plugins/rollup/public/crud_app/services/api.js:
```diff
@@ -30,6 +31,8 @@ export async function deleteJobs(jobIds) {
 
 export async function createJob(job) {
+  // Note that this will record an acion, even if the related request fails.
+  trackUserAction('create-rollup-job');
```
@yaronp68 The way this is implemented will result in this action being tracked even if the "create job" request fails, e.g. server-side validation rejects it. This will mean we'll see a discrepancy between the number of user actions to create a rollup job, and the number of rollup jobs themselves. Will this be a problem?

yaronp68 replied:
@cjcenizal I think that we are not trying to measure GUI failures vs. success and so having this "over-counting" is not desirable. However, If there's no good way to filter failures, it's not the end of the world as we will assume that usually the number represents the order of magnitude of using this feature (it doesn't really matter if there were 43 or 47 rollup jobs) and failures are marginal.

sebelga replied:
@cjcenizal why not indeed only track success requests?

```js
return await getHttp().put(`${apiPrefix}/create`, body)
    .then((response) => {
        trackUserAction('create-rollup-job');
        return response;
    })
```

cjcenizal commented on x-pack/plugins/rollup/public/crud_app/services/api.js:
```diff
@@ -30,6 +31,8 @@ export async function deleteJobs(jobIds) {
 
 export async function createJob(job) {
+  // Note that this will record an acion, even if the related request fails.
+  trackUserAction('create-rollup-job');
```
Note for reviewers: I think we should track all user actions by hitting this endpoint from the client, and not do any "proxied" user action tracking within CRUD API endpoints. I think this will a) make it easier to keep track of which actions are being tracked, and b) add greater fidelity to our data in the event that our API endpoints are consumed by other apps or third parties (e.g. command line or an automated tool). What does everyone else think?

epixa replied:
From a telemetry data perspective, why do we care if people created a rollup job by clicking a button in the UI versus using the API that powers that button? The benefit of putting the telemetry at the source of truth (the server) is that it represents all usage of the feature.

cjcenizal replied:
To provide more context, some of the actions we want to track aren't proxied by an API endpoint, e.g. clicking a particular tab or navigating somewhere. So I think the main benefit of keeping all tracking client-side will be that it's easier to keep track of what's being tracked, though maybe I'm overrating that benefit.

epixa replied:
I see.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1431/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1614/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1624/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1632/)

cjcenizal commented:
I passed this implementation by @jbudz and got his seal of approval. As long as we're tracking clicks and not mouse _movements_, the rate at which we're hitting the API should be fine. And the way the telemetry is gathered by the pertinent app (in this case, Rollups) is also correct.

sebelga commented:
I also wanted to ask, don't you think we should be doing this new feature in Typescript and following the new platform guidelines?

cjcenizal commented on src/legacy/core_plugins/user_action/server/routes/api/user_action.js:
```diff
@@ -0,0 +1,50 @@
+    path: '/api/user_action/{app}/{action_type}',
+    method: 'POST',
+    handler: async (request) => {
+      const { app, action_type: actionType } = request.params;
```
Haha silly me.

cjcenizal commented:
@sebelga I ❤️ both of the abstractions you've suggested! Currently, there's no way for an X-Pack plugin to relatively import code from an OSS plugin, because the location of the x-pack directory changes in production. This is being tracked in https://github.com/elastic/kibana/issues/31471. For now, we can just implement them at a shallower level, somewhere inside the `x-pack` directory.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2156/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2189/)

cjcenizal commented:
@sebelga I believe I've addressed your feedback. Could you take another look? I tried using the name format you suggested, but it looks like dashes and periods are automatically converted into underscores at some point. I did move the verb to the end though.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2195/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2196/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2224/)

sebelga commented on x-pack/plugins/rollup/public/crud_app/services/api.js:
```diff
@@ -31,7 +33,8 @@ export async function deleteJobs(jobIds) {
 export async function createJob(job) {
   const body = { job };
-  return await getHttp().put(`${apiPrefix}/create`, body);
+  const action = getHttp().put(`${apiPrefix}/create`, body);
```
To me "action" is synchronous. Here it is a "request", it can succeed or fail. It would then be clearer that is a promise in the `trackUserRequest()` method

bmcconaghy replied:
Agreed here.  I think we want to track success/failure.

sebelga commented on x-pack/plugins/rollup/public/crud_app/services/track_user_action.js:
```diff
@@ -0,0 +1,18 @@
+}
+
+export function trackUserRequest(action, actionType) {
+  // Only track successful actions.
```
There will be no use cases where we'll be interested in success/fail ratio on a user request? To me there are 3 telemetries on a request:

- action-requested
- action-succeed
- action-failed

@yaronp68 are we only interested in successes?

sebelga commented on x-pack/plugins/rollup/server/usage/collector.js:
```diff
@@ -180,6 +182,17 @@ export function registerRollupUsageCollector(server) {
 
+      const userActions = await fetchUserActions(server, 'rollup-job-wizard', USER_ACTIONS);
+
+      const userActionsByActionType = userActions.reduce((byActionType, userAction) => {
```
Why not wrap all this in a `byActionType(userActions)` helper?

sebelga commented:
Thanks for the changes @cjcenizal I still feel that we are _kindof_ addressing the problem by adding a helper to an x-pack server folder. If we think about the new platform where we will have clear plugin dependencies, the `fetchUserActions(), trackUserAction(), trackUserRequest()` will all be methods the plugin should return (out of a "registerApp" call).

Even with the limitation of paths that we currently have, I think we should have a solution for this problem so that the day we will migrate a plugin to the new platform we will just change the `registerApp()` call from one place to another.

What about using the  `common` folder?

x-pack
-- common
---- plugins
----- oss
------- user_actions
---------- lib
---------- index.js // registerApp() method here

I feel that if otherwise we'll have a bunch of apps copy / pasting (again) the `trackUserAction()` method when really it should just be a method call withouth having to know what the endpoint is for that (leaving it to the plugin) or if there is any more complex logic _before_ or _after_ the request.

yaronp68 commented:
@sebelga counting failures has a merit in telemetry but not sure how valuable it is assuming that it will be impossible to dig back to understand the cause unless it's a very obvious bug. I see more value with monitoring this for the monitoring GUI on the other hand

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2668/)

sebelga commented:
@cjcenizal yesterday we talked about folder structure during the sync, we decided it was better to export the `user_actions` plugin methods directly from `x-pack > common > user_actions`.
We can talk about it over zoom if u want.

cjcenizal commented:
@sebelga @bmcconaghy Ready for another review! Per discussion with Yaron I'm only tracking successful requests, since we won't be able to dig deeper and glean insights from failed request metrics.

I've also settled on what I believe to be the right level of abstraction by extracting out a `createUserActionUri` helper into `x-pack/common`. This helper ensures a constant format for hitting the user action API, while leaving it up to the consumer to determine how and when to hit it.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3033/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3052/)

cjcenizal commented:
@yaronp68 Can you confirm that you'll be able to create the visualizations you need using user action telemetry of this shape?

```json
{
  "rollups":{
    "user_actions":{
      "rollup_job_create":4,
      "rollup_job_delete":1,
      "rollup_job_start":1,
      "rollup_job_stop":1,
      "rollup_job_show_details_click":2,
      "rollup_job_detail_panel_summary_tab_click":1,
      "rollup_job_detail_panel_terms_tab_click":1,
      "rollup_job_detail_panel_histogram_tab_click":1,
      "rollup_job_detail_panel_metrics_tab_click":1,
      "rollup_job_detail_panel_json_tab_click":1
    }
  }
}
```

yaronp68 commented:
@cjcenizal 
Suggest to add UI in main field key (e.g. Rollups_UI) also speak to @Bamieh  and @pickypg about field naming conventions in telemetry 
I think the structure of the information is good and equivalent for other in telemetry

cjcenizal commented:
Thanks for the review @sebelga! I'm comfortable with this folder structure... let me explain my thoughts around this.

The split between OSS and X-Pack is necessary because there's no way to share code between the two until we have the new platform in place (i.e. you can't import OSS plugin code into an X-Pack plugin because the directory location changes in production, so relative imports will break). Only once the new platform is fully functional can we move this X-Pack code into OSS so it can be consumed by X-Pack plugins.

The split between the `x-pack/server` and `x-pack/common` makes sense to me because from conventions I've seen, one is for server-specific code and one is for code that is consumable by both client and server. I think it will be trivial to move `user_actions.ts` into a folder of the same name as the number of modules in this files grows and we need to scale the folder structure, but I'm also happy to make that change now.

cjcenizal commented:
I've talked with @pickypg about how to structure this data. ES has a soft limit of the number of fields set at 1000, and we're currently at just over 500. As we add more fields to track user actions, we'll rapidly approach and then exceed this limit. To avoid this problem, we've decided to store user action telemetry using a key-value format:

```
[
  { "key": "rollup_job_create", "value": 4 },
  { "key": "rollup_job_delete", "value": 7 }
]
```

The challenge with this is that it requires Kibana to support [nested fields](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html), which it currently does not. The ES team is working towards support for arbitrary objects containing key-value pairs that are not mapped but are still accessible queries. Until that point, we can reindex the index on-the-fly to generate the views we need, as a workaround.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3761/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3775/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3785/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/14720/)

sebelga commented:
This PR is reborn? 😊

rayafratkina commented:
@cjcenizal can you please add the DevDocs section to this pr?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/95/)
